pydantic/json_schema.py

```diff
@@ -513,20 +521,59 @@ def tagged_union_schema(self, schema: core_schema.TaggedUnionSchema) -> JsonSche
                     # may have been raised above, which would mean that the schema we want to reference won't be present
                     generated[str(k)] = generated[str(v)]
 
-        json_schema: JsonSchemaValue = {'oneOf': _deduplicate_schemas(generated.values())}
+        one_of_choices = _deduplicate_schemas(generated.values())
+        json_schema: JsonSchemaValue = {'oneOf': one_of_choices}
 
-        # This reflects the v1 behavior, but we may want to only include the discriminator based on dialect / etc.
-        if 'discriminator' in schema and isinstance(schema['discriminator'], str):
+        # This reflects the v1 behavior; TODO: we should make it possible to exclude OpenAPI stuff from the JSON schema
+        openapi_discriminator = self._extract_discriminator(schema, one_of_choices)
+        if openapi_discriminator is not None:
             json_schema['discriminator'] = {
-                # TODO: Need to handle the case where the discriminator field has an alias
-                #   Note: Weird things would happen if the discriminator had a different alias for different choices
-                #   (This wouldn't make sense in OpenAPI)
-                'propertyName': schema['discriminator'],
+                'propertyName': openapi_discriminator,
                 'mapping': {k: v.get('$ref', v) for k, v in generated.items()},
             }
 
         return json_schema
 
+    def _extract_discriminator(
+        self, schema: core_schema.TaggedUnionSchema, one_of_choices: list[_JsonDict]
+    ) -> str | None:
+        """
+        Extract a compatible OpenAPI discriminator from the schema and one_of choices that end up in the final schema.
+        """
+        openapi_discriminator: str | None = None
+        if 'discriminator' not in schema:
+            return None
+
+        if isinstance(schema['discriminator'], str):
+            return schema['discriminator']
+
+        if isinstance(schema['discriminator'], list):
+            # When an alias is used, the discriminator will be a list of single str lists, one for the attribute
+            # and one for the actual alias. This logic will work even if there is more than one possible attribute,
+            # and looks for whether a single alias choice is present as a documented property on all choices.
+            # If so, that property will be used as the OpenAPI discriminator.
+            for alias_path in schema['discriminator']:
+                if not isinstance(alias_path, list):
+                    break  # this means that the discriminator is not a list of alias paths
```

Comment:
surely in this case we should set `openapi_discriminator = alias_path`? at least if `alias_path` is a string?

Reply:
I realize I didn't handle the case of a single-element list properly. But if it was a list of more than one element it would be a path, which can't be handled. I will add handling of the one string element case. Let me know if you think there's more to it than that

Reply:
I was more thinking about `alias = ['foobar', [0, 'foo', 'bar']`, e.g. a choice of either `foobar` or a path, but maybe that's not allowed.

Reply:
My understanding was that that would be a malformed value of `alias` — I would expect that to look like `[['foobar'], [0, 'foo', 'bar']]`.

To be clear, it wouldn't be compatible with the type hint for discriminator: `str | list[str | int] | list[list[str | int]] | Callable[[Any], str | int | None]`

Reply:
type hint for validation_alias: `Union[str, List[Union[str, int]], List[List[Union[str, int]]]]` (same point stands)